8345678: Make Class.getModifiers() non-native.

coleenp opened the pull request:
Why isn't Class.getModifiers implemented in Java ?

coleenp commented on src/hotspot/share/classfile/javaClasses.cpp:
```diff
@@ -1500,7 +1500,8 @@ oop java_lang_Class::primitive_mirror(BasicType t) {
   macro(_reflectionData_offset,      k, "reflectionData",      java_lang_ref_SoftReference_signature, false); \
-  macro(_signers_offset,             k, "signers",             object_array_signature, false);
+  macro(_signers_offset,             k, "signers",             object_array_signature, false); \
+  macro(_modifiers_offset,           k, vmSymbols::modifiers_name(), int_signature,    false)
```
yes. it is needed.

ExE-Boss replied:
This is **C++**, so yes.
```suggestion
  macro(_modifiers_offset,           k, vmSymbols::modifiers_name(), int_signature,    false);
```

coleenp replied:
I see, there's a trailing semi somewhere so it compiles, but I added one in.

coleenp commented on src/java.base/share/classes/java/lang/Class.java:
```diff
@@ -1311,6 +1311,9 @@ private Class<?> elementType() {
     }
 
+    // Set by the JVM when creating the instance of this java.lang.Class
+    private transient int modifiers;
```
The JVM rearranges these fields so that's why I put it near the caller.

liach reviewed (commented):
The change to java.lang.Class looks good.

viktorklang-ora commented on src/java.base/share/classes/java/lang/Class.java:
```diff
@@ -1000,6 +1001,7 @@ public Module getModule() {
 
     private transient Object classData; // Set by VM
     private transient Object[] signers; // Read by VM, mutable
+    private final transient int modifiers;  // Set by the VM
```
@coleenp Could this field be @Stable, or does that only apply to `putfield`s?

liach replied:
I don't think this needs to be stable - finals in java.lang is trusted by the JIT compiler.

viktorklang-ora replied:
Yeah, I was just thinking whether something set from inside the VM which is marked @Stable is constant-folded :)

coleenp replied:
I don't think @Stable would hurt but final should provide the same guarantee.  It's set internally by the VM so there's no late setting. I don't know if this field implementation can constant fold in the case of Arrays which are (JVM_ACC_ABSTRACT | JVM_ACC_FINAL | JVM_ACC_PUBLIC).

DanHeidinga replied:
@viktorklang-ora `@Stable` is not about how the field was set, but about the JIT observing a  non-default value at compile time.  If it observes a non-default value, it can treat it as a compile time constant.

viktorklang-ora replied:
@DanHeidinga Great explanation, thank you!

dean-long replied:
If Class had other fields smaller than `int`, would be consider making this something like `char` to save space (allowing all the sub-word fields to be compacted)?

coleenp replied:
I thought of doing this since I made modifiers u2 in the Hotspot code just previously, but all the Java code refers to this as an int.  And I didn't see other fields to compact it with.  Maybe if access_flags are moved we could make them both char (not short since they're unsigned).  It feels weird to not have unsigned short to my C++ eyes.

liach replied:
From a Java perspective, using `char` for the field is completely fine; this field is only accessed via `getModifiers` and not set by Java code, so the automatic widening conversion can handle it all.

DanHeidinga commented on src/java.base/share/classes/java/lang/Class.java:
```diff
@@ -241,6 +241,7 @@ private Class(ClassLoader loader, Class<?> arrayComponentType) {
         // prevents future JIT optimizations from assuming this final field is null.
         classLoader = loader;
         componentType = arrayComponentType;
+        modifiers = 0;
```
The comment above about assigning a parameter to the field to prevent the JIT from assuming an incorrect default also should apply to the new `modifiers` field.  I think the constructor, which is never called, should also pass in a `dummyModifiers` value rather than using 0 directly

ExE-Boss commented on src/java.base/share/classes/java/lang/Class.java:
```diff
@@ -1001,6 +1002,9 @@ public Module getModule() {
     private transient Object classData; // Set by VM
     private transient Object[] signers; // Read by VM, mutable
 
+    @Stable
```
The `modifiers` field doesn’t need to be `@Stable`:
```suggestion
```

coleenp replied:
I now don't know whether we want @Stable here or not.

ExE-Boss commented on test/micro/org/openjdk/bench/java/lang/reflect/Clazz.java:
```diff
@@ -52,6 +55,34 @@ public Constructor getConstructor() throws NoSuchMethodException {
+     * @throws NoSuchMethodException
+     */
+    @Benchmark
+    public int getModifiers() throws NoSuchMethodException {
```
The only `Throwable`s that can be thrown by calling `Class::getModifiers()` are `Error`s (e.g.: `StackOverflowError`) and `RuntimeException`s (e.g.: `NullPointerException`):
```suggestion
    public int getModifiers() {
```

ExE-Boss commented on test/micro/org/openjdk/bench/java/lang/reflect/Clazz.java:
```diff
@@ -52,6 +55,34 @@ public Constructor getConstructor() throws NoSuchMethodException {
+
+    Clazz[] clazzArray = new Clazz[1];
+    @Benchmark
+    public int getAppArrayModifiers() throws NoSuchMethodException {
```
```suggestion
    public int getAppArrayModifiers() {
```

ExE-Boss commented on test/micro/org/openjdk/bench/java/lang/reflect/Clazz.java:
```diff
@@ -52,6 +55,34 @@ public Constructor getConstructor() throws NoSuchMethodException {
+     * @throws NoSuchMethodException
+     */
+    @Benchmark
+    public int getArrayModifiers() throws NoSuchMethodException {
```
```suggestion
    public int getArrayModifiers() {
```

coleenp commented:
/assign @iwanowww

liach commented:
Looking at #23396, we might need to filter this field too.

coleenp commented:
> Looking at https://github.com/openjdk/jdk/pull/23396, we might need to filter this field too.
Yes, I agree.

AlanBateman commented on src/hotspot/share/oops/arrayKlass.hpp:
```diff
@@ -1,5 +1,5 @@
 /*
- * Copyright (c) 1997, 2024, Oracle and/or its affiliates. All rights reserved.
+ * Copyright (c) 1997, 2025, Oracle and/or its affiliates. All rights reserved.
```
arrayKlass.hpp isn't changed, is this left over from a previous iteration?

coleenp replied:
yes, it was something that my copyright script thought I changed from merging some previous changes.

AlanBateman commented on src/java.base/share/classes/java/lang/Class.java:
```diff
@@ -236,11 +236,12 @@ private static void runtimeSetup() {
         // prevents future JIT optimizations from assuming this final field is null.
         classLoader = loader;
         componentType = arrayComponentType;
+        modifiers = dummyModifiers;
```
I realize this ctor isn't used but "dummyModifiers" looks very strange as parameter name when compared to the others, renaming it to something like "mods" would make it less confusing for anyone reading through this code.

coleenp replied:
I changed it to mods.  Thanks for the suggestion.

AlanBateman reviewed (approved):
Good cleanup.

coleenp reviewed (commented):
Thank you for your comments, Alan.

dean-long commented on test/micro/org/openjdk/bench/java/lang/reflect/Clazz.java:
```diff
@@ -52,6 +55,34 @@ public Constructor getConstructor() throws NoSuchMethodException {
+    @Benchmark
+    public int getAppArrayModifiers() {
+        return clazzArray.getClass().getModifiers();
+    }
```
I'm guessing this is the benchmarks that shows an extra load.  How about adding a benchmark that makes the Clazz[] final or @Stable, and see if that makes the extra load go away?

coleenp replied:
Name                      Cnt  Base   Error   Test   Error  Unit  Change
getAppArrayModifiers       30 0.923 ± 0.004  1.260 ± 0.001 ns/op   0.73x (p = 0.000*)
getAppArrayModifiersFinal  30 0.922 ± 0.000  1.260 ± 0.001 ns/op   0.73x (p = 0.000*)

No it doesn't really help.  There's still an extra load.

dean-long replied:
OK, if the extra load turns out to be a problem in the future, we could look into why the compilers are generating the load when the Class is known/constant.  If the old intrinsic was able to pull the constant out of the Klass, then surely we can do the same and pull the value from the Class field.

liach replied:
Does static final help here?

coleenp replied:
Yes.  Yes it does.

iwanowww replied:
Cases when a class mirror is a compile-time constant are already well-optimized. Non constant cases are the ones where missing optimization opportunities arise. 

In this particular case, C2 doesn't benefit from the observation that `Clazz[]` is a leaf type at runtime (no subclasses). Hence, a value loaded from a field typed as `Clazz[]` has exactly the same type and `clazzArray.getClass()` can be constant folded to `Clazz[].class`.  Rather than a common case, it feels more like a corner case. So, worth addressing as a follow-up enhancement.

Another scenario is a meet of 2 primitive array types (ends up as `bottom[]` in C2 type system), but I believe it hasn't been optimized before.

dean-long commented on src/hotspot/share/jfr/recorder/checkpoint/types/jfrTypeSet.cpp:
```diff
@@ -347,7 +347,7 @@ static void do_write_klass(JfrCheckpointWriter* writer, CldPtr cld, KlassPtr kla
   writer->write(mark_symbol(klass, leakp));
   writer->write(package_id(klass, leakp));
-  writer->write(klass->modifier_flags());
+  writer->write(klass->compute_modifier_flags());
```
Isn't this much more expensive than grabbing the value from the mirror, especially if we have to iterate over inner classes?

dean-long commented on src/hotspot/share/compiler/compileLog.cpp:
```diff
@@ -113,7 +113,7 @@ int CompileLog::identify(ciBaseObject* obj) {
         print(" unloaded='1'");
       } else {
-        print(" flags='%d'", klass->modifier_flags());
+        print(" flags='%d'", klass->access_flags());
```
There may be tools that parse the log file and get confused by this change.  Maybe we should also change the label from "flags" to "access flags".

dean-long commented on src/hotspot/share/oops/instanceKlass.hpp:
```diff
@@ -1125,7 +1125,7 @@ class InstanceKlass: public Klass {
 #endif
 
-  u2 compute_modifier_flags() const;
+  int compute_modifier_flags() const;
```
I don't see why this can't stay u2.

coleenp replied:
I had some compilation error for conversion that has disappeared into the either with u2, so I've restored them to u2.

dean-long commented on src/hotspot/share/opto/memnode.cpp:
```diff
@@ -2461,6 +2455,7 @@ const Type* LoadNode::klass_value_common(PhaseGVN* phase) const {
           // a primitive Class (e.g., int.class) has null for a klass field
           return TypePtr::NULL_PTR;
         }
+        // ???
```
I suspect that we still need this code to support intrinsics like LibraryCallKit::inline_native_classID() and maybe other users of this field, but the comment below no longer makes sense.

coleenp replied:
Thank you for noticing the ??? that I left in and the comment.

dean-long reviewed (approved):
Overall looks good to me.  Please ask @iwanowww to review compiler changes.

coleenp reviewed (commented):
Thank you for the detailed comments.

iwanowww reviewed (approved):
Looks good. (Except a left-over `???` in a comment.)

I very much like this cleanup. Migrating from Klass to Class simplifies compiler logic since there's no need to care about primitives at runtime.

Speaking of missing optimization opportunities (demonstrated by one microbenchmark), it looks like a corner case and can be addressed later.

coleenp commented:
Thank you Vladimir for encouraging me to continue this change.  I removed the ??? and hid the modifiers field for reflection as suggested in this PR.

coleenp commented:
I added some code to hide the Class.modifiers field and fixed the JVMCI test. Please re-review.

Also @iwanowww I think the intrinsic for isInterface can be removed and just be Java code like:

```
public boolean isInterface()  {
  return getModifiers().isInterface();
}
```

iwanowww commented:
> I think the intrinsic for isInterface can be removed

Good point. Moreover, it seems most of intrinsics on Class queries can be replaced with a flag bit check on the mirror. (Do we have 16 unused bits in Class::modifiers after this change?)

coleenp commented:
> Good point. Moreover, it seems most of intrinsics on Class queries can be replaced with a flag bit check on the mirror. (Do we have 16 unused bits in Class::modifiers after this change?)

Yes, I think so.  isArray and isPrimitive definitely.  We could first change the modifiers field to "char" because that's its size and then have two booleans for each of these.

liach commented:
Making `isArray` and `isPrimitive` Java-based is going to be helpful for the interpreter performance of these methods in early bootstrap. :+1:

AlanBateman reviewed (approved):
No more comments from me.

mur47x111 reviewed (approved):
JVMCI change looks good to me

coleenp commented:
Thank you for the reviews Yudi, Alan, Chen, Vladimir and Dean, and the help and comments with the various pieces of this.
/integrate
